lib/content_encoding.c

```diff
@@ -132,65 +133,85 @@ inflate_stream(struct connectdata *conn, contenc_writer *writer)
   /* Dynamically allocate a buffer for decompression because it's uncommonly
      large to hold on the stack */
   decomp = malloc(DSIZ);
-  if(decomp == NULL) {
+  if(decomp == NULL)
     return exit_zlib(conn, z, &zp->zlib_init, CURLE_OUT_OF_MEMORY);
-  }
 
   /* because the buffer size is fixed, iteratively decompress and transfer to
-     the client via client_write. */
-  for(;;) {
-    if(z->avail_in == 0) {
-      free(decomp);
-      return result;
+     the client via downstream_write function. */
+  while(!result && z->avail_in) {
+    /* Check state. */
+    switch(zp->zlib_init) {
+    case ZLIB_INIT:
+    case ZLIB_INFLATING:
+    case ZLIB_INIT_GZIP:
+    case ZLIB_GZIP_INFLATING:
+      break;
+    default:
+      result = exit_zlib(conn, z, &zp->zlib_init, CURLE_WRITE_ERROR);
+      continue;
     }
 
     /* (re)set buffer for decompressed output for every iteration */
     z->next_out = (Bytef *) decomp;
     z->avail_out = DSIZ;
 
     status = inflate(z, Z_SYNC_FLUSH);
-    if(status == Z_OK || status == Z_STREAM_END) {
-      allow_restart = 0;
+    /* Flush output data if some. */
+    if(z->avail_out != DSIZ) {
+      zp->zlib_init = started;      /* Data started. */
       result = Curl_unencode_write(conn, writer->downstream, decomp,
                                    DSIZ - z->avail_out);
-      /* if !CURLE_OK, clean up, return */
       if(result) {
-        free(decomp);
-        return exit_zlib(conn, z, &zp->zlib_init, result);
-      }
-
-      /* Done? clean up, return */
-      if(status == Z_STREAM_END) {
-        free(decomp);
-        return exit_zlib(conn, z, &zp->zlib_init, result);
+        exit_zlib(conn, z, &zp->zlib_init, result);
+        break;
       }
-
-      /* Done with these bytes, exit */
-
-      /* status is always Z_OK at this point! */
-      continue;
     }
-    else if(allow_restart && status == Z_DATA_ERROR) {
+    /* Dispatch by inflate() status. */
+    switch(status) {
+    case Z_OK:
+      break;
+    case Z_STREAM_END:
+      /* There may be 8 gzip trailer bytes to read.
+         Use z->avail_out as a counter. */
+      z->avail_out = started == ZLIB_GZIP_INFLATING? 8: 0;
+      if(z->avail_in < z->avail_out) {
+        z->avail_out -= z->avail_in;
+        zp->zlib_init = ZLIB_GZIP_TRAILER;
+      }
+      else {
+        if(z->avail_in > z->avail_out)
+          result = CURLE_WRITE_ERROR;
+        result = exit_zlib(conn, z, &zp->zlib_init, result);
+      }
+      z->avail_in = 0;
+      break;
```

Comment:
Is this check necessary and did you account in these changes for concatenated gzip streams, I've read that is [possible](https://groups.google.com/d/msg/comp.compression/UuGWViGLPOY/I0EpcCmPwpIJ).

Reply:
This check is not "necessary": it just allows detecting trailing bytes (after the stream), while the previous code silently ignored the trailing bytes in buffer and restartied a possibly desynced stream at the next call if some. See https://github.com/curl/curl/blob/462f3cac34c09b9c28b449258e71dc37171dc604/lib/content_encoding.c#L128
Returning an error for unparsable trailing bytes is probably cleaner than silently discarding them.

> did you account in these changes for concatenated gzip streams, I've read that is possible.

No, and I think this was never supported in curl: see the above comment. In any case, use of zlib  < 1.2.0.4 cannot allow it if we don't process the trailer explicitly.

The link to Mark Adler's comment is about gzip **files** and this is primarily a mean to support gzip archives with several files. Strictly speaking, you're right about concatenation (see RFC 2616, 3.5). In practice, I don't know if it is also intended for HTTP streams, if some server can generate it or if some client supports it and how (cannot find something about it on internet). Functionally speaking, concatenation only makes sense if curl does not inflate (thus outputting a real gzip file).

Supporting concatenation could be a new feature. TBD since it implies a loss of structure data.

Reply:
> Returning an error for unparsable trailing bytes is probably cleaner than silently discarding them.

Probably but there's a lot of wackiness out there.